### PR TITLE
[reporelaypy] Update checkout with hook instead of polling

### PR DESCRIPTION
**Pull request by JonWBedard**
#### 8aa39443d8057d743b10d7929f50dcd6ff72b7e2
<pre>
[reporelaypy] Update checkout with hook instead of polling
<a href="https://bugs.webkit.org/show_bug.cgi?id=234243">https://bugs.webkit.org/show_bug.cgi?id=234243</a>
&lt;rdar://problem/86413065 &gt;

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/libraries/reporelaypy/reporelaypy/__init__.py: Bump version,
export HookProcessor and HookReceiver.
* Tools/Scripts/libraries/reporelaypy/reporelaypy/checkout.py:
(Checkout.update_for): If a branch is new, we need to track it.
* Tools/Scripts/libraries/reporelaypy/reporelaypy/database.py:
(Database.__init__): Use self.host and self.password.
* Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py: Added.
(HookProcessor): Class to process received hooks.
(HookReceiver): Class to receive hooks and queue them for processing.
* Tools/Scripts/libraries/reporelaypy/reporelaypy/tests/hooks_unittest.py: Added.
(HooksUnittest):
(HooksUnittest.setUp):
(HooksUnittest.test_receive):
(HooksUnittest.test_invalid):
(HooksUnittest.test_process):
(HooksUnittest.test_hmac):
(HooksUnittest.test_invalid_hmac):
* Tools/Scripts/libraries/reporelaypy/reporelaypy/webserver.py: Add hook routes,
if hooks are enabled.
* Tools/Scripts/libraries/reporelaypy/run: Allow caller to enable hooks.
* Tools/Scripts/libraries/reporelaypy/setup.py: Bump version.
</pre>

### CODE REVIEW

File: Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+class HookProcessor(object):
+    INBOUND_KEY = 'inbound-hooks'
+    WORKER_HOOKS = 'worker-hooks'
+    TYPES = ('push',)
+
+    @classmethod
+    def is_valid(cls, type, data):
+        if type not in cls.TYPES or not isinstance(data, dict):
+            return False
+        if type == 'push' and data.get('ref'):
```
Comment (dewei-zhu) [inline Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py]: Nit: can be simplified as `return type == 'push' and data.get('ref')`

File: Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    def process_hooks(self):
+        with self.database.lock(name='lock_{}'.format(self.INBOUND_KEY), timeout=60):
+            for key in self.database.scan_iter('{}:*'.format(self.INBOUND_KEY)):
+                data = json.loads(self.database.get(string_utils.decode(key)))
+                self.process_hook(type=data.get('type'), data=data.get('data'))
+
+                encoded = json.dumps(data)
+                for i in range(self.num_workers):
+                    digest = hashlib.md5()
+                    digest.update(string_utils.encode(encoded))
```
Comment (dewei-zhu) [inline Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py]: Do we need to calculate md5 for encoded multiple times? we can just calculate hexdigest once out of this for loop.
Also worth mentioning md5 is not safe, but might be OK for getting a unique key.

**Review by dewei-zhu — APPROVED** [review]
r=me with comments.

File: Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py (hunk 2 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/reporelaypy/reporelaypy/hooks.py]: Just looking for a unique key.

**Comment by JonWBedard** [comment]
Testing a mention for automation: @dewei-zhu

**Comment by JonWBedard** [comment]
Landed https://commits.webkit.org/245242@main (d1b61724f4a8366c7058d2fe238a2d2eccf42c48)

**Pull request closed without merging**